staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go

```diff
@@ -251,6 +265,16 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 				// client disconnect.
 				return
 			}
+
+			// If event is BOOKMARK, skip to flush to client
+			if event.Type == watch.Bookmark {
+				*bookmarkEvent = *outEvent
```

Comment:
I just skimmed the code quickly, so I might missed something, but I think this is simply incorrect.

If there was some other even send in the meantime after this handler received bookmark but before watch timed out, we may simply send a bookmark with lower RV, which breaks the watch invariants.

i'm fairly sceptical that we can really make this feature work like that in a safe and performant enough way.

Reply:
> If there was some other even send in the meantime after this handler received bookmark but before watch timed out, we may simply send a bookmark with lower RV, which breaks the watch invariants.

Please see the follow codes:

```go
			// If event is not BOOKMARK, set bookmarkEvent to nil, because it's not the latest resource version
			bookmarkEvent = nil
```
---

> i'm fairly sceptical that we can really make this feature work like that in a safe and performant enough way.

I did some simple tests, it works fine. Maybe I missed some cases.

Reply:
aah ok - I missed that. but I have comments below.